# TP 2 : récursivité et conditions

Dans tout le TP, il est interdit d'utiliser des boucles (pas de `for` ni `while`).

Rappel : pensez à tester toutes vos fonctions sur des exemples.

## Petites questions

1. Définir une fonction `divise` telle que `divise a b` renvoie `true` si  `a` divise `b`, `false` sinon. Il est interdit d'utiliser `if`.
2. Définir une fonction récursive `somme` telle que `somme n` renvoie $\sum_{k=1}^n k^2$.
3. Définir une fonction récursive `u` telle que `u n` renvoie $u_n$ définie par :
$$u_0 = 42$$
$$u_{n} = 3\sqrt{u_{n - 1}} + 2$$
4. Définir une fonction récursive `v` telle que `v a n` renvoie $v_n$ définie par (on fera attention à n'utiliser qu'un appel récursif) :
$$v_0 = a$$
$$v_{n + 1} = \frac{1}{2}(v_n + \frac{a}{v_n})$$
5. Calculer `v a n` pour plusieurs valeurs de `a` et de `n` : que peut-on conjecturer sur la limite de $v_n$ ? Optionnel : le démontrer.  
6. Soient $a$ et $b$ deux entiers et $q, r$ le quotient et reste de la division euclidienne de $a$ par $b$ ($a = bq + r$). En utilisant le fait que $PGCD(a, b)$ = $PGCD(b, r)$, écrire une fonction `euclide` telle que `euclide a b` renvoie le PGCD de $a$ et $b$.

In [2]:
let divise a b = b mod a=0;;
let rec somme ?(k=1) n = if k=n then k*k else k*k + somme  ~k:(k+1) n;;
let rec u n = if n=0 then 42. else 3.*.(u (n-1))**0.5+.2.;;
let rec v a n = if n=0 then a else let j=v a (n-1) in 0.5*.(j+.a/.j);;

val divise : int -> int -> bool = <fun>


val somme : ?k:int -> int -> int = <fun>


val u : int -> float = <fun>


val v : float -> int -> float = <fun>


Il semblerait que quand n tende vers l'infini v a n tende vers racine de a. Par ailleurs on remarque que si 
v n = sqrt(a) alors v (n+1) = 1/2(sqrt(a)+a/sqrt(a)) = sqrt(a) <=> v n(+1) = v(n).

In [4]:
let rec euclide a b = if b = 0 then a else euclide b (a mod b);;
euclide 420 3003;

val euclide : int -> int -> int = <fun>


- : int = 21


## Exponentiation rapide

1. Écrire une fonction récursive `puissance` naïve (c'est à dire très simple) telle que `puissance a n` renvoie $a^n$. Combien effectue t-elle de multiplications (en fonction de n) ?
2. Écrire une fonction récursive `exp_rapide` pour calculer $a^n$, en utilisant les relations suivantes :
$$ 
\begin{cases} 
a^n = (a^{\frac{n}{2}})^2 ~~~~~~~~~\text{si }n\text{ est pair}\\
a^{n} = a \times (a^{\frac{n-1}{2}})^2 ~~~~~\text{sinon}
\end{cases}
$$
**Attention** : pour mettre au carré, utiliser une variable plutôt que faire 2 appels récursifs, ce qui augmenterait beaucoup le nombre d'opérations.  
**Remarque** : On montrera plus tard que cette 2ème version demande de l'ordre de $\ln(n)$ multiplications seulement.

In [1]:
let rec puissancenaive a n = if n=1 then a else a * puissancenaive a (n-1);;
let rec exp_rapide a n = if n = 1 then a else if divise 2 n then let j = exp_rapide a (n/2) in j*j else let j = exp_rapide a ((n-1)/2) in a*j*j;; 

val puissancenaive : int -> int -> int = <fun>


error: compile_error

# Accumulateur

On a vu dans le cours sur la récursivité (avec l'exemple de la suite de Fibonacci) qu'un accumulateur est un argument que l'on ajoute à une fonction pour calculer sa valeur de retour.  
1. Écrire une fonction `fact` telle que `fact acc n` renvoie `n`!, en utilisant `acc` comme accumulateur. Voici à quoi va ressemble `fact` :
```ocaml
let rec fact acc n =
    if n = 0 then acc  (* on renvoie l'accumulateur qui contient le résultat *)
    else fact ... (* appel récursif en modifiant l'accumulateur *)
```
2. En utilisant `fact` et l'application partielle de fonction, définir `f : int -> int` renvoyant la factoielle d'un entier.

**Remarque** : le but ici est juste de vous entraîner à savoir utiliser un accumulateur, qui sont parfois utiles (comme pour la fonction `fibo` du cours). En DS ou concours on évitera d'utiliser un accumulateur lorsqu'il y en a pas besoin (comme pour la fonction `fact`...), car cela rend le code plus compliqué.  

In [17]:
let rec fact ?(acc=1) n = if n = 0 then acc else fact ~acc:(acc*n) (n-1);;
fact 4

val fact : ?acc:int -> int -> int = <fun>


- : int = 24


## Temps de vol de la suite de Syracuse

La suite de Syracuse d'un entier $a$ est définie par :  
$$u_0 = a$$
$$u_{n+1} =
\begin{cases} 
     \frac{u_n}{2}, \text{si } n \text{ est pair}\\
    3u_n + 1, \text{sinon}\\
\end{cases}$$

Le temps de vol de $(u_n)_n$ est le plus petit entier $t$ tel que $u_t = 1$.  

1. Écrire une fonction récursive `temps_vol` telle que `temps_vol a` renvoie le temps de vol de $(u_n)_n$ (où $u_0 = a$).

In [22]:
let rec temps_vol ?(u=0) a = if a=1 then u else if divise 2 a then temps_vol ~u:(u+1) (a/2) else temps_vol ~u:(u+1) (3*a+1);;
temps_vol 27

val temps_vol : ?u:int -> int -> int = <fun>


- : int = 111


## Fonction mystérieuse

Tester la fonction suivante, conjecturer un théorème et prouvez-le par récurrence.

In [39]:
let rec f n =
    if n > 100 then n - 10 
    else f(f (n + 11)) ;;

val f : int -> int = <fun>


Montrons que P(n) ("si n>100 alors f(n) renvoie n-10, si n<=100 alors f(n) renvoie 91") est vraie pour tout n appartenant à N.
Par définition P(n) est vrai pour tout n>100.
Faisons un récurence descendante forte à partir de n = 100.
Initialisation si n = 100 alors le code renvoie f(f(111)) = f(101) = 91.

Récurrence supposons p(n) vraie pour tout n appartenant à [[100,k]], k fixe.
Montrons que p(k-1) est vraie. f(k-1) = f(f(k+10)).
Si k+10>100, alors f(f(k+10)) = f(k). D'après l'hypothese de récurrence, f(k)=91 donc f(k-1)=91. Donc p(k-1) est vraie.
Si k+10 <=100 alors f(f(k+10)) = f(91) d'après l'hypothèse de récurrence. Donc f(k-1)=91.
Donc P(n) est vrai pour tout n appartenant a N.


let rec f x =
if x = 0 then 2
else let j = f (x-1) in j - j*(g (x-1))
and
g x =
if x=0 then 3
else let k = g (x-1) in k + k*(f (x-1));;

f 3## Fonctions mutuellement récursives

Il est possible de définir simultanément deux fonctions `f` et `g`, dépendant l'une de l'autre avec `and` :
```ocaml
let rec f x = ... and g y = ... in ...
```

Écrire deux fonctions `u` et `v` permettant de calculer le $nième$ terme des suites définies par :
$$u_0 = 2$$
$$v_0 = 3$$
$$u_{n + 1} = u_n - u_n v_n$$
$$v_{n + 1} = v_n + u_n v_n$$

In [3]:
let rec f x =
if x = 0 then 2
else let j = f (x-1) in j - j*(g (x-1))
and
g x =
if x=0 then 3
else let k = g (x-1) in k + k*(f (x-1));;

f 3

val f : int -> int = <fun>
val g : int -> int = <fun>


- : int = 896


## Retour sur les tours de Hanoi

**Remarque** : le problème des tours de Hanoi a déjà été vu en stage, mais on le refait ici en OCaml. En outre, il est utile de réactiver sa mémoire pour se souvenir des méthodes et il m'arrivera pendant l'année de redonner des exercices déjà posés.

![](../img/hanoi.png)

$n$ disques sont posés sur la tige à gauche. L'objectif est de déplacer tous les disques sur la tige à droite :

![](../img/hanoi2.png)

Règles du jeu :

- On ne peut déplacer qu'un disque à la fois (celui tout en haut), sur une autre tige.
- Il est interdit de poser un disque sur un autre plus petit.

Exemple de premier déplacement valide :

![](../img/hanoi3.png)

On souhaite écrire une fonction récursive `hanoi` telle que `hanoi n tige1 tige2` affiche une suite de déplacements (avec des `print_int`) permettant de déplacer $n$ disques depuis `tige1` vers `tige2`. On supposera que les tiges sont numérotées 0, 1, 2 (de gauche à droite).

1. Supposons que vous sachiez déplacer $n-1$ disques d'une tige à une autre. Comment déplacer $n$ disques d'une tige à une autre ?
2. Écrire `hanoi`.

Si je sais déplacer n-1 disques d'une tige A à une tige B pour déplacer n disques d'une disque à une autre, il suffit de déplacer n-1 disques de la tige A à la tige C puis de déplacer le dernier disque de la tige A à la tige B puis de redéplacer les n-1 disques de la tige C à la tige B.

In [10]:
let rec hanoi n tige1 tige2 = 
if n=1 
then (print_int(tige1);print_string "=>";print_int(tige2);print_string("\n"))
else (hanoi (n-1) tige1 (6-(tige1 + tige2)); hanoi 1 tige1 tige2; hanoi (n-1)(6-(tige1 + tige2)) tige2);;

hanoi 3 1 2;
print_newline()

val hanoi : int -> int -> int -> unit = <fun>


1=>2
1=>3
2=>3
1=>2
3=>1
3=>2
1=>2



- : unit = ()


interrupt: intterupt

In [7]:
print_int(5);

print_newline()
print_int(5);
print_newline()

55


- : unit = ()
